### PR TITLE
Add capabilities to get and render manpage output in the Web UI

### PR DESCRIPTION
**Pull request by rdmark**
- class method to get arbitrary file contents
- new web UI endpoint to fetch the contents of rascsi man pages (rendered txt files)
- Link to manpage in page footer
- Link to new wiki page for image type docs
- Other UI tweaks

### CODE REVIEW

**Comment by uweseimet** [comment]
@rdmark A better user experience might be to generate HTML from the raw manpage data. See https://linux.die.net/man/1/man2html. Not sure whether this is better, though.

File: python/web/src/web.py [hunk 1]
```diff
@@ -487,6 +487,29 @@ def show_diskinfo():
...
+        return response(
+            template="manpage.html",
+            app=app,
+            manpage=manpage,
+            version=server_info["version"],
+            )
+
+    return response(
+        error=True,
+        message=_("An error occurred when accessing man page: %(error)s", error=logs)
```
Comment (nucleogenic) [inline python/web/src/web.py:509]: `logs` is not defined in this function.

Comment (rdmark) [reply python/web/src/web.py]: Good catch. The dangers of copy-pasting.

**Comment by nucleogenic** [comment]
One thing to consider with this change set is that the standalone web UI install does not install the manpages, so we get an error. These are installed as part of the `make install` of RaSCSI as far as I can see. Perhaps they can be copied to the correct location when option 11 (standalone web) is chosen?

File: python/common/src/rascsi/sys_cmds.py [hunk 2]
```diff
@@ -181,3 +181,18 @@ def get_diskinfo(file_path):
...
         return process.returncode, process.stderr.decode("utf-8")
+
+    @staticmethod
+    def get_manpage(app):
+        """
+        Takes (str) app, the name of the application to fetch the man page for.
+        Returns either the man output, or the stderr output.
+        """
+        process = run(
+                ["man", app],
```
Comment (nucleogenic) [inline python/common/src/rascsi/sys_cmds.py:192]: Could you add `man-db` to the apt install line in `rascsi-web/Dockerfile`?

**Comment by rdmark** [comment]
> One thing to consider with this change set is that the standalone web UI install does not install the manpages, so we get an error. These are installed as part of the `make install` of RaSCSI as far as I can see. Perhaps they can be copied to the correct location when option 11 (standalone web) is chosen?

That's a fair point. However, then you install the Web UI stand-alone, wouldn't you already have installed RaSCSI proper earlier? Without the RaSCSI backend running, the Web UI will error out when you try to access it.

I'm not opposed to the improvement, but it seems a bit redundant at a glance. :)

**Comment by rdmark** [comment]
manerr

FWIW the error handling is pretty graceful :)

File: python/common/src/rascsi/sys_cmds.py (hunk 2 above)
Comment (rdmark) [reply python/common/src/rascsi/sys_cmds.py:192]: Sure! What does it do?

**Comment by nucleogenic** [comment]
As a workaround, could we try an approach like this, where we give the path to the docs instead?

```
man <base dir>/doc/rascsi.1 
```

**Comment by rdmark** [comment]
> As a workaround, could we try an approach like this, where we give the path to the docs instead?
> 
> ```
> man <base dir>/doc/rascsi.1 
> ```

Good idea! Although I'd say the rendered txt is even better to display. Added some code to do this.

File: python/common/src/rascsi/sys_cmds.py (hunk 2 above)
Comment (nucleogenic) [reply python/common/src/rascsi/sys_cmds.py:192]: It just adds `man` which isn't included in the minimal Debian images.

File: python/web/src/web.py [hunk 3]
```diff
@@ -487,6 +488,39 @@ def show_diskinfo():
...
+            template="manpage.html",
+            app="rascsi",
+            manpage=manpage,
+            version=server_info["version"],
+            )
+
+    return response(
+        error=True,
+        message=_("An error occurred when accessing man page: %(error)s", error=manpage)
+    )
```
Comment (nucleogenic) [inline python/web/src/web.py:521]: Aha! I wasn't actually aware these existed.

In this case, can the whole thing be reduced to something like (untested!) the following?

```
app_whitelist = ["rascsi", "rasctl", "rasdump", "scsimon"]

app = request.args.get("app")

if app not in app_whitelist:
    return response(
        error=True,
        message=_(".... not in whitelist error ....")
    )

file_path = f"{WEB_DIR}/../../../doc/{app}_man_page.txt"

try:
    with open(file_path) as file:
        return response(
            template="manpage.html",
            app=app,
            manpage=file.read(),
            version=server_info["version"],
        )
except Exception as error:
    return response(
        error=True,
        message=_("An error occurred when accessing man page: %(error)s", error=error)
    )
```

File: python/common/src/rascsi/sys_cmds.py [hunk 4]
```diff
@@ -181,3 +181,33 @@ def get_diskinfo(file_path):
...
+                capture_output=True,
+                )
+        if process.returncode == 0:
+            return process.returncode, process.stdout.decode("utf-8")
+
+        return process.returncode, process.stderr.decode("utf-8")
+
+    @staticmethod
+    def get_filecontents(file_path):
```
Comment (nucleogenic) [inline python/common/src/rascsi/sys_cmds.py:201]: I think this method is redundant if we can use `open()` instead and get nicer exception-based error handling.

**Review by nucleogenic — COMMENTED** [review]
Rather than using calls to external programs, I think we should utilise the pre-formatted RaSCSI documentation text files, if we can. I wasn't aware they existed, so using `man` for the formatting seemed like the best idea when I first commented (sorry!)

I do not think there is much of a use case for a general purpose `man` wrapper, but I could be wrong on that.

The code example I gave could be enhanced to remove the two warning lines, e.g. by using `file.readlines()` instead of `file.read()` and manipulating the list.

File: python/common/src/rascsi/sys_cmds.py (hunk 4 above)
Comment (rdmark) [reply python/common/src/rascsi/sys_cmds.py:186]: Rewrote this to use open() instead

File: python/common/src/rascsi/sys_cmds.py (hunk 2 above)
Comment (rdmark) [reply python/common/src/rascsi/sys_cmds.py]: Removed the dependency on man, so reverted this change.

File: python/web/src/web.py (hunk 3 above)
Comment (rdmark) [reply python/web/src/web.py]: Thanks for writing the code for me. ;)
I adopted this approach, and simplified it further to skip the man step, and read the txt file always.

**Comment by rdmark** [comment]
> Rather than using calls to external programs, I think we should utilise the pre-formatted RaSCSI documentation text files, if we can. I wasn't aware they existed, so using `man` for the formatting seemed like the best idea when I first commented (sorry!)
> 
> I do not think there is much of a use case for a general purpose `man` wrapper, but I could be wrong on that.
> 
> The code example I gave could be enhanced to remove the two warning lines, e.g. by using `file.readlines()` instead of `file.read()` and manipulating the list.

I went a slightly different route, keeping the get_filecontents() class method generic, and then stripping out the header lines in the show_manpage() endpoint. It's not the most memory efficient way, but it doesn't really matter for this use case I think.

**Review by nucleogenic — APPROVED** [review]

**Pull request merged**